File: python/p92.py
# initial time 13.7s
# change get_next to use % and / 7.97


def get_next(n):
    sum = 0
    while n > 0:
        d = n % 10
        sum += d * d
        n = n // 10
    return sum


def ends89(n):
    while n != 1 and n != 89:
        next = get_next(n)
        n = next
    return n == 89


count = 0
for i in range(1, 10000000):
    if ends89(i):
        count += 1
print(count)
if count != 8581146:
    print("AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA")

File: go/p92/p92.go
// fun fact this takes 0.29 s in go
// in python it takes 7.97
package main

import "fmt"

func getNext(n int) int {
	sum := 0
	for n > 0 {
		d := n % 10
		sum += d * d
		n = n / 10
	}
	return sum
}

func ends89(n int) bool {
	for n != 1 && n != 89 {
		n = getNext(n)
	}
	return n == 89
}

func main() {
	count := 0
	for i := 1; i < 10000000; i++ {
		if ends89(i) {
			count += 1
		}
	}
	fmt.Println(count)
}
